vulnerabilities/overflow-underflow.md

```diff
@@ -1,73 +1,150 @@
+
+```solidity
+interface IDexPair {
+    function getSwapQuoteUint128(uint128 amountToSwap) external view returns(uint128);
+}
+
+contract DexPair is IDexPair {
+    function getSwapQuoteUint128(uint128 amountToSwap) external view returns(uint128 outputTokens) {
+        assembly {
+            outputTokens := add(amountToSwap, 1)
+            if lt(outputTokens, amountToSwap) { revert(0, 0) }
+        }
+    }
+}
+```
+
+- The problem is that the `add` opcode always produces a 256-bit result. For `uint128` maximum value (`type(uint128).max`), this will not overflow in the 256-bit space but will overflow when treated as `uint128`
+    
+- The inline assembly overflow check fails because the EVM uses 256-bit arithmetic, while the function is supposed to handle 128-bit values. This discrepancy allows an overflow to evade detection, causing the function to return incorrect results without reverting. Specifically, adding 1 to the maximum uint128 value (2^128 - 1) results in an overflow to 0 instead of triggering the overflow detection.
+
 
 ### Use of unchecked code block:
-Performing arithmetic operations inside the unchecked block saves a lot of gas because it omits several checks and opcodes. But, some of these opcodes are used in default arithmetic operations in 0.8 to check for underflow/overflow.
+Performing arithmetic operations inside the unchecked block saves a lot of gas because it omits several checks and opcodes. But some of these opcodes are used in default arithmetic operations in 0.8 to check for underflow/overflow.
 
 ```solidity
 uint8 public a = 255;
 
 function uncheck() public{
 
   unchecked {
-      a++;  // overflow and reset to 0 without reverting
+      a++;      // overflow and reset to 0 without reverting
   }
 
 }
 ```
 The unchecked code block is only recommended if you are sure that there is no possible way for the arithmetic to overflow or underflow.
 
+
+### Mitigation Strategies
+
+1. Consider using `addmod` for the subtle `uint128` integer overflow. The `addmod` operation confines the result within the `uint128` range.
+
+```solidity
+function getSwapQuoteUint128(uint128 amountToSwap) external view returns(uint128 outputTokens) {
+    assembly {
+        outputTokens := addmod(amountToSwap, 1, 340282366920938463463374607431768211455)
+        if lt(outputTokens, amountToSwap) { revert(0, 0) }
+    }
+}
+```
+
+2. Consider performing a post-assembly check for the subtle `uint128` integer overflow. Perform an overflow check outside YUL using normal Solidity to ensure correct behavior for `uint128` as it will compare the 128-bit values.
+
+```solidity
+function getSwapQuoteUint128(uint128 amountToSwap) external view returns(uint128 outputTokens) {
+    assembly {
+        outputTokens := add(amountToSwap, 1)
+    }
+    require(outputTokens >= amountToSwap, "Overflow detected!");
+}
+```
+
+3. Add a manual check. The below check for example ensures that if the result is less than the input, it indicates an overflow.
```

Comment:
Is this not the same as number 2?

Reply:
We can probably simply the mitigation section to just:
- Revert if an over/underflow will occur (will happen regardless with safemath/v>0.8)
- Don't use unchecked blocks unless over/underflow is impossible